### ISSUE TITLE
New Account Ledger vs Account Main Balance

### ISSUE THREAD

**Issue by kdjstudios** [kdjstudios/SABillingGitlab#877]
In GitLab by @smasih on May 7, 2018, 01:10

Query and list the accounts which have different numbers on Ledger and main balance on account level.

After listing the accounts, we need to come up with a plan to resolve this calculation issue.

**Comment by kdjstudios** [comment]
In GitLab by @smasih on May 7, 2018, 04:11

@tim.traylor @kdjstudios 

Here is the list of accounts with different acc and ledger balance ( **Current ledger is not including Pending staged Fees** ).

Total accounts: 15006    
Faulty Accounts: 4005  


[ledger_vs_acc_balance.csv](/uploads/89d96a1c5f05f1858e048eac1847f4b6/ledger_vs_acc_balance.csv)


We think, next thing to do would be to identify the scenarios which leads to miscalculation.

@tim.traylor May we please know your input as well.

**Comment by kdjstudios** [comment]
In GitLab by @tim.traylor on May 7, 2018, 06:01

@thomas.gentry Hi Thomas, I think that some of these differences may be historical data that was prior to sites converting and being on SAB.  We need to identify the original conversion date and make sure we do not calculate ledger info prior to that date.

Please gather the original conversion date for each of our sites that are still active and add the file to this issue ticket. Please complete this today so the team can work on it tomorrow.

@smasih Sumeet, we will also need to add a new field to the Site data to hold this value.

Thx

**Comment by kdjstudios** [comment]
In GitLab by @pchaudhary on May 7, 2018, 06:55

Marking Primary for #876 as both of the tickets serving the same purpose.

**Comment by kdjstudios** [comment]
In GitLab by @thomas.gentry on May 7, 2018, 08:43

Attached is the conversion date for each of our sites sorted by date.

[SAB_sites_original_conversion_date.xlsx](/uploads/ec4759bf7e070b804775da1595b76ae0/SAB_sites_original_conversion_date.xlsx)

**Comment by kdjstudios** [comment]
In GitLab by @pchaudhary on May 8, 2018, 07:40

Hi [Kyle] ( @kdjstudios ) & [Tim] ( @tim.traylor ),   

We have created a new account_vs_ledger list based on the conversion date provided by Thomas.  


[Ledger_vs_balance.csv](/uploads/aa589e4dd5ce8971b1015d8ec459421a/Ledger_vs_balance.csv)

**Comment by kdjstudios** [comment]
In GitLab by @tim.traylor on May 8, 2018, 09:14

The next step will be for the Aavaz team to do the calculation on 3 or 4 accounts and see if a pattern is visible.  Also, Tim commented that he thinks it is odd that almost every ledger amount calc is a negative value.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on May 15, 2018, 09:44

@smasih @tim.traylor @pchaudhary @thomas.gentry 

During the converstation today, it was mentioned that this is due to the account ledger calculating items from prior to the Orginal conversion date.

Please be sure that the account ledger does not calculate any items on or before the site's conversion date. http://gitlab.aavaz.biz/AnswerNet/SABilling/issues/877#note_19712

**Comment by kdjstudios** [comment]
In GitLab by @smasih on May 17, 2018, 05:58

@kdjstudios @thomas.gentry @tim.traylor 
@pchaudhary @abakshi 

Account ledger vs Main account balance comparison has **4005** accounts. [ledger_vs_account_balance.csv](/uploads/3e36b3c8d2db43d1dfbaf0bbc80a8f63/ledger_after_conversion_dates.csv)  

After implying Tim's suggestion, which was to reject transactions before site's conversion date  the comparison dropped to **445** accounts. [ledger_after_conversion_dates_merge2.csv](/uploads/eb72f6a2d3a3bdf193540df318ca107e/ledger_after_conversion_dates_merge2.csv)

This data we submitted was from merge which we updated two days ago with production data.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on May 17, 2018, 09:47

We will need to identity the accounts that have priority issues. Such as the externals first,; and then we will address the internals.

Please filter the list by active and terminated; and also accounts over and under $100.00 variance.

If there is a way to programmatically tell the issue of accounts rather then reviewing on a one by one bases please let us know.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on May 17, 2018, 10:38

@smasih 

I just want to confirm in this ticket, that we are not only rejecting transactions before the sites conversion date, but also the ones on that date.

**Comment by kdjstudios** [comment]
In GitLab by @pchaudhary on May 17, 2018, 23:57

@kdjstudios 

> I just want to confirm in this ticket, that we are not only rejecting transactions before the sites conversion date but also the ones on that date.

* Yes, for sure. We are also rejecting the transactions held on the conversion date.

Thanks.

**Comment by kdjstudios** [comment]
In GitLab by @smasih on May 18, 2018, 04:38

> We will need to identity the accounts that have priority issues. Such as the externals first,; and then we will address the internals.

We don't have conversion dates for external clients so, this data is only for answernet internals.



> Please filter the list by active and terminated; and also accounts over and under $100.00 variance.

We divided further groups like `less than $100`, `between $100 to $500`, `between $500 to $1000`, and  `greater than $1000`.

[ledger_after_conversion_dates_merge4.csv](/uploads/7c0e219ab4e957665bbdcedc8b9f85ab/ledger_after_conversion_dates_merge4.csv)



> If there is a way to programmatically tell the issue of accounts rather then reviewing on a one by one bases please let us know.

Unfortunately, there is not, we have to identify scenarios which are creating the issue and then find a way to resolve it.

**Comment by kdjstudios** [comment]
In GitLab by @tim.traylor on May 18, 2018, 06:02

@smasih [P.S.E._.G] and [OSU_HAVNER_EYE_INSTITUTE] are the two accounts to start researching.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on May 18, 2018, 08:23

@smasih @pchaudhary @tim.traylor @thomas.gentry 

I do apologize as I may be confused, but thought we resolved the conversion dates and that was not the full solution (Thus why we are having to look into each acount); so YES we still need to do this for EXTERNALS first!

Please confirm that there is no issues with the externals.

**Comment by kdjstudios** [comment]
In GitLab by @pchaudhary on May 22, 2018, 07:03

Hi [tim] ( @tim.traylor ) & [Kyle] ( @kdjstudios ),   


> [P.S.E._.G] and [OSU_HAVNER_EYE_INSTITUTE] are the two accounts to start researching.

* We have dug into the accounts & below are our findings respectively to each account - 


1.  P.S.E.&.G - The difference of $ - 2318.4 is due to some payments which were not reflected to any invoice. We can see that there is one payment & an invoice of the account which does have the same creation date which leads us to another scenario & we tested it again after changing the conversion date of **Atlanta** from `21^st Feb,2014` to `26^th Mar, 2014`.

1.  OSU HAVENER EYE INSTITUTE/DEPT - The same scenario as above in which some payments have not reflected in an invoice. We can see that there is two payments & an invoice of the account which does have the same creation date which leads us to another scenario & we tested it again after changing the conversion date of **Columbus,OH** from `6^th Mar,2015` to `1^st Apr, 2015`.




*   Please find the list of accounts having the account ledger difference for Hosted.  [ledger_after_conversion_dates_merge.csv](/uploads/732871e1137462d8d32023e1804b7fa2/ledger_after_conversion_dates_merge.csv)

**Note** - All of the above calculations was done in the Merge server.


Thanks

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on May 22, 2018, 09:05

@tim.traylor @smasih @pchaudhary @thomas.gentry 

Okay so once we remove the 250+ NEIT accounts; since that site is not active, we are down to about 20 accounts. [External_Account_Ledger_Error.csv](/uploads/89254f96e2e0f12405443cf518aedd14/External_Account_Ledger_Error.csv)

After filtering the spreadsheet, it looks like Keener Lynchburg has three active accounts that have over $500 difference, I would believe those should be the ones we first look into, not the internal accounts. (Per Gary)

**Comment by kdjstudios** [comment]
In GitLab by @smasih on May 23, 2018, 05:17

@kdjstudios @tim.traylor 

Worked on following two accounts: (following are the only accouns for Laser and Towne having ledger balance mismatch)

| Laser Answering Service | Pro Vista Solutions, Inc | PV0002 |
|-------------------------|--------------------------|--------|
| Towne Answering Service | G & R Mechanical, Inc.   | 3731   |


Account Pro Vista had its 1st invoice with credit balance but not recorded in sab (more investigation required)

G&R mechanical is having an incorrect value calculated on invoice and in DB 

on Invoice it says `$2180.03` but after calculation it comes to be `2121.7`

in DB this value is `1030.42`. The difference between these two balance is the difference b/w ledger and main account amount.

In short the calculation for this invoice is incorrect.![SA_Hosted_and_Managed_Services](/uploads/5ab41b97bc4668c4eaa74f74b7abca64/SA_Hosted_and_Managed_Services.png)

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on May 23, 2018, 08:24

@smasih @pchaudhary @rchauhan @abakshi @tim.traylor @thomas.gentry 

I am confused we specifically during the meeting with Gary were requested to look into the external accounts first and those that had the greatest impact. So we provide you with the list and inform you to look into the three active accounts that have over a $500 difference. 

But it appears that you completely disregarded the request and did no research on the accounts that were mentioned. However, you look into two other accounts in complete opposition of the request??

**Comment by kdjstudios** [comment]
In GitLab by @smasih on May 24, 2018, 24:35

@kdjstudios 

I am failing to see that how we disregarded the request, the request was to fix the issue and we are trying our best to do that. Although, we might have modified the strategy a bit. Believe me, we just  wanted to eliminate the two external sites from the equation (which were having only one account) so that we can solely focus on keener accounts.

**Comment by kdjstudios** [comment]
In GitLab by @smasih on May 24, 2018, 05:00

@kdjstudios 

Just to provide the background, I missed on your comment so, was not aware that we have been asked to work on the keener accounts. Never, intended to disregard the request.

We discussed our findings with Tim (yesterday during our morning huddle), and he wants us to keep digging.


Coming back to issue, we found that most the accounts in "keener lynchburg" having the issue are those accounts which were cloned from Keener Cali site to Keener Lynchburg. (Including active accounts above $500 difference)

Accounts were not transferred but copied over inorder to keep the history of Keener Cali, but the previous balance was not transferred over to the new accounts.

Need to discuss with @tim.traylor on how to handle this scenario. 


**Here is the updated list with an additional _ISSUE_ column** [External_Account_Ledger_Error_with_issue.csv](/uploads/32ad6d36c07c99845666537b5d955e20/External_Account_Ledger_Error_with_issue.csv)

**Comment by kdjstudios** [comment]
In GitLab by @smasih on May 24, 2018, 07:48

Hi [Kyle] ( @kdjstudios ) & [Tim] ( @tim.traylor ), 


We ran the comparison, with changed conversion dates (suggested by us) & ledger calculations for all of the faulty accounts from those particular sites got resolved.

Thanks.

**Comment by kdjstudios** [comment]
In GitLab by @pchaudhary on May 25, 2018, 04:49

Hi [Kyle] (@kdjstudios ) & [Tim] ( @tim.traylor ),   

### Internal

We have done a more investigation on the issue & found that there are some more sites which should have some different conversion dates because there are some transactions with the equal timestamp.

Based on our investigation, we have changed the Site Conversion date for  *Allentown, Fair Oaks, Orlando,  Atlanta, Chattanooga & Columbus* & the accounts which do have still remained the wrong ledger calculation dropped to **8 accounts only**.

*  Please find the updated list of changed conversion date. [SAB_sites_original_conversion_date.xlsx](/uploads/9070b951a608fe52cb81752d3d77249e/SAB_sites_original_conversion_date.xlsx)  

*  Please find the list of 8 accounts having the wrong ledger balance. [ledger_after_conversion_dates_merge.csv](/uploads/3740b62df7a4081427f2847b83ac5534/ledger_after_conversion_dates_merge.csv)


Thanks.

**Comment by kdjstudios** [comment]
In GitLab by @pchaudhary on May 25, 2018, 05:51

Hi [Kyle] ( @kdjstudios ) & [Tim] ( @tim.traylor ),  

#### External

We have done a more investigation for the external accounts also & found that most of the wrong ledger balance accounts are due to the transfer from Keener Cali to Keener Lynchburg which was also stated by Sumeet on his [comment] (http://gitlab.aavaz.biz/AnswerNet/SABilling/issues/877#note_20521).

We have investigated for the remaining accounts also & below are our findings: - 

1.  Keener Communications - ReMax Commonwealth (6266) -  
  *  3 check payments were done by the same check number & 2 of them are having the same timestamp.  
![image](/uploads/ee72b8f9f22d13fa093be3d28d71c547/image.png)

1.  Keener Communications - Remax Commonwealth - Alverser (1875) -   
  *  2 check payments were done by the same check number & both of them are having the same timestamp.  
![image](/uploads/99ac85e99b6450ada80554127b7d1129/image.png)

1.  Keener Lynchburg - Seven Hills Surgical Assoc / CMG 1911 (6744) - 
  *  2 check payments were done by the same check number & both of them are having the same timestamp.  
![image](/uploads/ae1a9f2eb6b780f05e4c4167bf3728f8/image.png)

1.  Keener Lynchburg - Altavista Alarm (6784) - 
  *  The calculation of last invoice (KNRVA-14493) of the account does not reflect the previous balance.



* **Note** - Please find the updated list of externals accounts[External_Account_Ledger_Error_with_issue.csv](/uploads/8eb3fc8e90081add5d819808f164195f/External_Account_Ledger_Error_with_issue.csv) ledgers faulty accounts.  


Thanks.

**Comment by kdjstudios** [comment]
In GitLab by @smasih on May 26, 2018, 01:51

@kdjstudios @tim.traylor 

Released **New Account Ledger** code (both servers) which will honour site's conversion date.

On external there are no conversion dates.

Currently on internal conversion dates are those which we got from the [file](http://gitlab.aavaz.biz/AnswerNet/SABilling/issues/877#note_19712) posted by Thomas.

@pchaudhary have suggested different conversion dates for some sites, but didn't get chance to discuss it by Tim. So, once discussed we will make the required DB edits.

>> Note: Just to be clear, there will be no code changes required after today for this particular issue but DB edits will be required.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on May 29, 2018, 09:11

@thomas.gentry Is going to compile the adjusted conversion dates and then the team will update Ledger accordingly.

**Comment by kdjstudios** [comment]
In GitLab by @tim.traylor on Jun 4, 2018, 06:23

@kdjstudios Hi Kyle - I just want to check that you saw the file Piyush posted on 5/25 that has the 17 accounts for the externals that need to be corrected.  As with previous issues, we'll want to run it by them and get their permission before fixing.  Thx

**Comment by kdjstudios** [comment]
In GitLab by @thomas.gentry on Jun 4, 2018, 08:19

Here is an updated list of sites and adjusted conversion dates.

Each date is the end_date of the first master billing cycle that follows the "created_at" date of the first billing cycle.

In simpler terms each date should be the end_date of the first billing_cycle that didn't use historic data.

[Adjusted_SAB_Sites_original_conversion_date.xlsx](/uploads/b4f9d142684d640114fb67cbc1ba7447/Adjusted_SAB_Sites_original_conversion_date.xlsx)

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Jun 4, 2018, 09:24

I wrote to each external client about the faulty account ledgers and will create new tickets for their responses accordingly.

**Comment by kdjstudios** [comment]
In GitLab by @pchaudhary on Jun 5, 2018, 07:16

Hi [Thomas] ( @thomas.gentry ), [Kyle] ( @kdjstudios ) & [Tim] ( @tim.traylor ),  

We have copied the production server data to the Merge server & changed the conversion date as per the file provided by Thomas.
We ran the script to find the accounts having the wrong ledger balances & here is the file having them. [ledger_after_conversion_dates_merge.csv](/uploads/04f19650ff1783584dc9e9849f0b4299/ledger_after_conversion_dates_merge.csv) 
*  There are total `95` accounts remaining to adjust ledger.

* We also would like to mention that the coversion date provided by Thomas is not exactly what we suggested on our [comment] (http://gitlab.aavaz.biz/AnswerNet/SABilling/issues/877#note_20590) after which there were only `8` accounts remained.

Thanks.

**Comment by kdjstudios** [comment]
In GitLab by @pchaudhary on Jun 7, 2018, 06:26

#925 - Keener   
#926 - Laser   
#924 - Towne - **Completed**

**Comment by kdjstudios** [comment]
In GitLab by @tim.traylor on Jun 15, 2018, 06:21

@kdjstudios Hi Kyle, can you please follow up with the externals and see if we can get approval for the guys to make the changes?  It would be nice if we can put this issue to bed.  Thx

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Jun 15, 2018, 13:11

@tim.traylor 

Thank you I have followed up on both Keener and Laser tickets accordingly.

**Comment by kdjstudios** [comment]
In GitLab by @tim.traylor on Jun 19, 2018, 09:27

@pchaudhary Piyush, can you pls upload the conversion dates you calculated so Thomas can compare them to his dates?

**Comment by kdjstudios** [comment]
In GitLab by @pchaudhary on Jun 20, 2018, 24:14

Hi [Tim] (@tim.traylor ) & [Thomas] (@thomas.gentry ),   

I have already uploaded the conversion dates in one of my comment. [Click here] (http://gitlab.aavaz.biz/AnswerNet/SABilling/issues/877#note_20590) .   

Please check.

Thanks.

**Comment by kdjstudios** [comment]
In GitLab by @tim.traylor on Jun 20, 2018, 06:22

@thomas.gentry Thomas, Piyush had looked at transactions on sites and suggested his dates based on where there was a large volume of records all with the same/very close timestamps.

**Comment by kdjstudios** [comment]
In GitLab by @smasih on Jun 22, 2018, 08:20

@kdjstudios 

no partial release on this ticket for current cycle so, moving to next milestone.

**Comment by kdjstudios** [comment]
In GitLab by @pchaudhary on Jul 5, 2018, 02:05

Hi [Kyle] ( @kdjstudios ) & [Thomas] ( @thomas.gentry ),   

We are still waiting for the response of Thomas to proceed further.

**Comment by kdjstudios** [comment]
In GitLab by @pchaudhary on Jul 5, 2018, 02:05

assigned to @pchaudhary

**Comment by kdjstudios** [comment]
In GitLab by @pchaudhary on Oct 10, 2018, 06:49

Hi [Kyle] ( @kdjstudios ),   

Can we close this one?

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Oct 10, 2018, 09:01

@pchaudhary 

Sorry as it appears its been multiple months since we last updated this ticket and I hardly recall what happened and what needs to happen. Judging from the previous posts above there are issues with the account main balance and the account ledger balance. It also appears that we were able to update these to match for the externals; but I do not see that these were updated for the internals? May you please confirm.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Nov 14, 2018, 15:06

@pchaudhary @amishra @smasih @tim.traylor 

May we please confirm that status of this ticket. Thank you.

**Comment by kdjstudios** [comment]
In GitLab by @pchaudhary on Nov 16, 2018, 07:04

@kdjstudios 


This can be closed as the account ledger is fixed now.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Nov 16, 2018, 12:28

Thank you marking for close

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Nov 23, 2018, 11:45

closed
